### PR TITLE
MYST-20 Communication dialog request/Response DTO models

### PR DESCRIPTION
**Pull request by Waldz**

### CODE REVIEW

**Comment by Waldz** [comment]
Improvements done, review appreciated

**Comment by tadovas** [comment]
Is it possible to extract hardcoded nats server address/port to options somehow? Or at least to declare inner constant in nats package. Now "127.0.0.1" is lying around everywhere.

File: communication/codec_bytes_test.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
+package communication
+
+import (
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestCodecBytesInterface(t *testing.T) {
+	var _ Codec = NewCodecBytes()
```
Comment (ignasbernotas) [inline communication/codec_bytes_test.go]: This test does nothing?

File: communication/codec_json_test.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+type customPayload struct {
+	Field int
+}
+
+func TestCodecJSONInterface(t *testing.T) {
+	var _ Codec = NewCodecJSON()
```
Comment (ignasbernotas) [inline communication/codec_json_test.go]: Same as above, does it do anything?

File: communication/nats_discovery/address.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	return &NatsAddress{
+		servers: []string{
+			fmt.Sprintf("nats://%s:%d", server, port),
+		},
+		topic: topic,
+	}
+}
+
+func NewAddressForIdentity(identity dto_discovery.Identity) *NatsAddress {
+	return NewAddress("127.0.0.1", 4222, string(identity))
```
Comment (ignasbernotas) [inline communication/nats_discovery/address.go]: Is it ok to hardcode "127.0.0.1" + port here?

Comment (ignasbernotas) [reply communication/nats_discovery/address.go]: Maybe worth changing into a variable we could modify during compilation?

File: communication/codec_bytes_test.go (hunk 1 above)
Comment (Waldz) [reply communication/codec_bytes_test.go]: This test check if struct implements wanted interface

File: communication/codec_bytes.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+		return []byte(payload), nil
+	}
+
+	return []byte{}, fmt.Errorf("Cant pack payload: %#v", payloadPtr)
+}
+
+func (codec *codecBytes) Unpack(data []byte, payloadPtr interface{}) error {
+	switch payload := payloadPtr.(type) {
+	case *[]byte:
+		*payload = data
```
[the remark this answers is not in the document]
Comment (Waldz) [reply communication/codec_bytes.go]: It's casting.
But we dont have byte streams yet, so I dont want to polish them without real case.

File: communication/nats_discovery/address.go (hunk 3 above)
Comment (Waldz) [reply communication/nats_discovery/address.go]: It's the same how it was before, we expect queue server to run next to node&client.
We have a ticket to make it injectable

File: communication/codec_bytes.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Waldz) [reply communication/codec_bytes.go]: What was the idea of story:
- you can write custom providers/consumers
- providers/consumers will provide DTOs of requests/responses
- DTOs can be any structs, what are able to serialise to JSON

**Pull request merged**